pkg/p2p/server_client_integration_test.go

```diff
@@ -0,0 +1,380 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/phayes/freeport"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/ticdc/pkg/security"
+	"github.com/pingcap/ticdc/proto/p2p"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+// read only
+var clientConfig4Testing = &MessageClientConfig{
+	SendChannelSize:         128,
+	BatchSendInterval:       time.Millisecond * 200,
+	MaxBatchCount:           128,
+	MaxBatchBytes:           8192,
+	RetryRateLimitPerSecond: 10.0, // using 10.0 instead of 1.0 to accelerate testing
+	DialTimeout:             time.Second * 3,
+}
+
+type serverConfigOpt = func(config *MessageServerConfig)
+
+func newServerForIntegrationTesting(t *testing.T, serverID string, configOpts ...serverConfigOpt) (server *MessageServer, addr string, cancel func()) {
+	port := freeport.GetPort()
+	addr = fmt.Sprintf("127.0.0.1:%d", port)
+	lis, err := net.Listen("tcp", addr)
+	require.NoError(t, err)
+
+	var opts []grpc.ServerOption
+	grpcServer := grpc.NewServer(opts...)
+
+	serverConfig := *defaultServerConfig4Testing
+	for _, opt := range configOpts {
+		opt(&serverConfig)
+	}
+
+	server = NewMessageServer(serverID, &serverConfig)
+	p2p.RegisterCDCPeerToPeerServer(grpcServer, server)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_ = grpcServer.Serve(lis)
+	}()
+
+	cancel = func() {
+		grpcServer.Stop()
+		wg.Wait()
+	}
+	return
+}
+
+func runP2PIntegrationTest(ctx context.Context, t *testing.T, size int, numTopics int) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
L163 already calls `cancel()`

Reply:
Added the deferred cancel here in case there is a panic mid-way.